### PR TITLE
Add challenge and project creation and editing forms

### LINKED ISSUE #16: Project creation form
We need a form for users to create projects.

### LINKED ISSUE #17: Challenge creation form
We need a form for users to create challenges.

### LINKED ISSUE #19: Challenge participation flow
There should be an option to "Participate" in a challenge which will allow a user to select a project which already exists and work on that, or create their own.

### PR DESCRIPTION
**Pull request by fluix-dev**
TODO:
 - [x] Button for creation
 - [x] Button for editing
 - [x] Permission checks for editing
 - [ ] ~~Only display certain types of tags for addition~~ See #36 
 - [x] Fix Chosen tag dropdown

And in a separate PR:
 - [ ] Convert profile editing to `base_form.html`
 - [ ] Normalize URL naming conventions
 - [ ] Alphabetize Python imports
 - [ ] Make description mandatory

Fixes #16 
Fixes #17 
Fixes #19

### CODE REVIEW

**Comment by fluix-dev** [comment]
@dulldesk or @skyflaren could one of you take a look at the CSS issue.

File: app/templates/base_form.html [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      {% if form.errors %}
+        <p class="text-danger">There were errors in your form. Please correct them.</p>
+      {% endif %}
+      {% for error in form.non_field_errors() %}
+        <p class="text-danger">{{ error }}</p>
+      {% endfor %}
+    </div>
+    <form action="." method="POST">
+      {% csrf_token %}
+      <div class="frame__body">
```
Comment (dulldesk) [inline app/templates/base_form.html]:
```suggestion
      <div class="frame__body" style="overflow-y:initial">
```
Prevents the tags selection dropdown from overflowing the div and making it scroll.

**Comment by dulldesk** [comment]
Is it possible to force the challenge start/end fields to render as time inputs?

**Comment by dulldesk** [comment]
Forms need to redirect to a URL upon completion. 
> ImproperlyConfigured at /challenge/create/
No URL to redirect to.  Either provide a url or define a get_absolute_url method on the Model.

Perhaps they could redirect to their creation?

**Comment by dulldesk** [comment]
9f90968 removes the padding that was added for the icon in `explore` but is unnecessary in inputs that don't add an icon in the input bar (i.e. in the challenge/project creation/edit forms).

**Comment by dulldesk** [comment]
Perhaps add a note somewhere in the form that required fields are marked blue.

File: app/templates/base_form.html (hunk 1 above)
Comment (fluix-dev) [reply app/templates/base_form.html]: What happens instead?

**Edit**: Just tested, looks really nice :+1:

**Comment by fluix-dev** [comment]
> Perhaps add a note somewhere in the form that required fields are marked blue.

I was thinking of adding `(Optional)` to the time fields. Some others are _technically_ optional but we want them filled out.

**Edit**: I don't want to do this anymore. `help_text`s shouldn't have this kind of information. Changing color also aligns with what Django's admin does.

File: app/templates/base_form.html (hunk 1 above)
Comment (dulldesk) [reply app/templates/base_form.html]: The drop down list overlays on top of whatever happens to be below it (in this case, the footer). No extra space is created.

**Comment by dulldesk** [comment]
That could work. Would the label colouring still be needed?

**Comment by fluix-dev** [comment]
> That could work. Would the label colouring still be needed?

Nah, not really.

**Edit**: I'm a liar, see above.

**Comment by dulldesk** [comment]
We should probably add a warning somewhere that if you remove your account as a `creator`, then you lose `creator` permissions. (Or, alternately, add a modal warning if a user is trying to do such a thing). 

Should we also automatically add the user as a `creator` for a project/challenge they are trying to create, e.g. by adding the user account to the `creators` input field (and they can choose to remove it)?

**Comment by dulldesk** [comment]
Will permissions for creating challenges be restricted?

**Comment by dulldesk** [comment]
Upon creating a project for a challenge this error is thrown: 
> NameError at /project/create/<challenge pk>/
> name 'get_success_url' is not defined

**Comment by fluix-dev** [comment]
> We should probably add a warning somewhere that if you remove your account as a creator, then you lose creator permissions. (Or, alternately, add a modal warning if a user is trying to do such a thing).

Yeah I'll add this into the `help_text`. You actually can't remove all creators, so at least one person will always be able to edit.

> Should we also automatically add the user as a creator for a project/challenge they are trying to create, e.g. by adding the user account to the creators input field (and they can choose to remove it)?

I think the easiest way to do this would be through a bit of JS, so yeah I'll do it.

> Will permissions for creating challenges be restricted?

I'm not sure, but it's probably a good idea. I initially thought everyone can until they are "muted" or do something silly, but limiting challenge creation to only specific people in the first place is probably a good idea.

**Comment by fluix-dev** [comment]
@Gongsta what do you think about challenge creation permissions? Should everyone be able to create a challenge or only people that contact us and get the permission?

**Comment by fluix-dev** [comment]
Alright, I think I fixed everything.

**Comment by dulldesk** [comment]
Should we add a notice if a challenge has been closed, e.g. in the place of the `Create Project` (for a challenge) button? The user may not realise the timestamps' presence, especially as they are given equal emphasis to (e.g.) tags.

**Review by dulldesk — CHANGES_REQUESTED** [review]
This issue does not seem to have been resolved yet. https://github.com/KuriusMTL/HackItForward/pull/33#issuecomment-694245133

**Comment by fluix-dev** [comment]
> This issue does not seem to have been resolved yet. #33 (comment)

Fixed, and improved in general.

File: app/models.py [hunk 2]
```diff
@@ -170,6 +173,20 @@ def short_creators(self):
...
+
+    def get_absolute_url(self):
+        return reverse("challenge", args=[self.pk])
+
+    def get_edit_url(self):
+        return reverse("challenge_edit", args=[self.pk])
+
+    def is_open(self):
+        return self.start <= timezone.now() <= self.end
+
```
Comment (dulldesk) [inline app/models.py]: An error is thrown when the challenge timestamps are blank

File: app/models.py [hunk 3]
```diff
@@ -167,6 +168,15 @@ class Challenge(models.Model):
         help_text="Cover image of this challenge.",
     )
 
+    def clean(self):
+        super().clean()
+        if self.start and not self.end:
+            raise ValidationError("An end date must be provided if you have a start date.")
+        if not self.start and self.end:
+            raise ValidationError("An start date must be provided if you have an end date.")
```
Comment (dulldesk) [inline app/models.py]:
```suggestion
            raise ValidationError("A start date must be provided if you have an end date.")
```
Typo

**Comment by dulldesk** [comment]
[diff](https://github.com/KuriusMTL/HackItForward/pull/33/files#diff-b5099d1f0106025f31781ccbe0a59d13R163-R166)

`context["time_labels"]` may display null timestamps, should anything be done about this?

**Comment by fluix-dev** [comment]
This should have been covered by https://github.com/KuriusMTL/HackItForward/issues/11

**Comment by dulldesk** [comment]
Timestamps were not blankable then but that is valid.

**Comment by fluix-dev** [comment]
Fair, then it should have been https://github.com/KuriusMTL/HackItForward/pull/40

**Comment by fluix-dev** [comment]
It's super simple, but not related to this, so let's just commit it after.

**Review by dulldesk — APPROVED** [review]

**Pull request merged**